### PR TITLE
Release/v0.6.1

### PR DESCRIPTION
**Pull request by mosufy**
Bumped to v0.6.1

Key additions:
- Utils/db.selectPaginate
- Utils/isEmail
- Utils/prepSQLInsertParams
- Utils/prepSQSUpdateParams
- Utils/validateFields
- Utils/isDecimal

Others:
- Middlewares/httpMiddleware now extracts JWT.sub from Authorization Header to event.auth.sub if available

### CODE REVIEW

File: src/services/AuroraDbService.js [hunk 1]
```diff
@@ -45,6 +47,30 @@ export default class AuroraDbService {
...
+  ) {
+    let paginator;
+    if (typeof total === 'number') {
+      paginator = new LengthAwarePaginator(
+        this,
+        sql,
+        sqlParams,
+        perPage,
+        currentPage,
+        total
```
[the remark this answers is not in the document]
Comment (jpaylaga) [reply src/services/AuroraDbService.js:65]: I can now confirm that it is all good as the `=== true` is actually not necessary.

File: src/utils/isDecimal.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+export default number => {
```
Comment (chocnut) [inline src/utils/isDecimal.js:1]: We can improve this by 

`export default () => typeof number !== 'string' && number.toString().indexOf('.') !== -1`

File: src/utils/isEmail.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import isEmpty from './isEmpty';
+import LesgoException from '../exceptions/LesgoException';
+
+export default email => {
+  const pattern = /^[\w\-\.\+]+\@[a-zA-Z0-9\.\-]+\.[a-zA-z0-9]{2,5}$/; // eslint-disable-line no-useless-escape
+
+  if (isEmpty(email)) {
```
Comment (chocnut) [inline src/utils/isEmail.js:7]: Since you're using typescript might be better to be explicit as to what is the type of email argument here. By doing that you don't need to use isEmpty

**Review by chocnut — APPROVED** [review]
a couple of small comments but it doesn't really block the release.

File: src/services/AuroraDbService.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpaylaga) [reply src/services/AuroraDbService.js:65]: They are actually important if we want to use LengthAwarePaginator on the auroradb with it without supplying the total number and use the internally available logic to get the total count of the query.

In the current code right now, the LengthAwarePaginator will only be used if the total params is supplied by a number, which will represent the total count, which can be derived somewhere, say, somebody has created a more efficient query to get the total number.

If we want the AuroraDbService to use LengthAwarePaginator  and  make it calculate the total count of the query automatically, we have to supply the boolean true , then the condition inside will determine if LengthAwarePaginator should be supplied with a number or null.

[the remark this answers is not in the document]
Comment (jpaylaga) [reply src/services/AuroraDbService.js:65]: My apologies for the late realization.

File: src/services/__tests__/LengthAwarePaginator.spec.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  db.select.mockClear();
+});
+
+const FILE = 'Services/pagination/LengthAwarePaginator';
+
+describe('test LengthAwarePaginator instantiate', () => {
+  it('should not throw exception when instantiating', async () => {
+    const paginator = new LengthAwarePaginator(
+      db,
+      'SELECT * FROM tests',
```
Comment (daison12006013) [inline src/services/__tests__/LengthAwarePaginator.spec.js:22]: I didn't notice that we should pass a query statement inside the length aware paginator, how about an interface of collection to be passed, where we have another class that decorates a `handle()` or `collection()`

So this class will never be dependent to call a real database, we can just create a class that returns a collection (easy way of mocking it)

Just a suggestion that we can improve soon.

File: src/services/AuroraDbService.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mosufy) [reply src/services/AuroraDbService.js:65]: If the LengthAwarePaginator requires it to be a number, than it should do the check within the LengthAwarePaginator and not on the function that is calling it. This is to prevent other functions from misusing it as well. Which I believe is already handled and will throw an Error if it is not null nor a number.

If this function requires a total to be a null or a number, than it needs the same validation as well. But this is not the explicit requirement of this selectPaginate function correct?

[the remark this answers is not in the document]
Comment (jpaylaga) [reply src/services/AuroraDbService.js:65]: Yes, that is not the explicit requirement of the `selectPaginate` function. The additional condition is specific for the `selectPaginate` function because that determines whether to create a `Paginator` or a `LengthAwarePaginator`.

File: src/utils/__mocks__/db.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        { ...mockDataFirstItem },
+        { ...mockData },
+        { ...mockData },
+        { ...mockData },
+        { ...mockDataLastItem },
+        { ...mockData },
+      ]);
+    }
+
+    if (sql.startsWith('SELECT * FROM total_tests')) {
```
Comment (daison12006013) [inline src/utils/__mocks__/db.js:52]: Ohh, if we're mocking a data on this way, I think in the long run we can just use sqlite and populate it on runtime, rather than checking if the statement `startsWith`

**Review by daison12006013 — DISMISSED** [review]
Looks good to me 👍 , just have a minor suggestions.

File: src/services/__tests__/LengthAwarePaginator.spec.js (hunk 4 above)
Comment (chocnut) [reply src/services/__tests__/LengthAwarePaginator.spec.js:22]: plus one to this. clean code

File: src/services/pagination/LengthAwarePaginator.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+   * Constructor
+   *
+   * @param db
+   * @param sql
+   * @param sqlParams
+   * @param perPage
+   * @param currentPage
+   * @param total
+   */
+  constructor(db, sql, sqlParams, perPage = 10, currentPage = 1, total = null) {
```
Comment (chocnut) [inline src/services/pagination/LengthAwarePaginator.js:17]: I just noticed this. But I'm not sure if you guys have this rule. Most of the time we apply to max-params for any given class or function. If it's more than 3 I'll convert this to be an object instead. The reason being is that you might miss the order of params even tho some of them are optional.

File: src/services/pagination/LengthAwarePaginator.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    if (total !== null && typeof total !== 'number') {
```
Comment (chocnut) [inline src/services/pagination/LengthAwarePaginator.js:18]: I don't get it, total param is optional however you don't want this to be null? why not explicitly check only for a number type instead?

File: src/services/pagination/LengthAwarePaginator.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  /**
+   * Total items in all pages.
+   *
+   * @returns {null|number}
+   */
+  async total() {
+    if (this.totalProp === null) {
+      this.totalProp = await this.countTotalItems();
```
Comment (chocnut) [inline src/services/pagination/LengthAwarePaginator.js:46]: Can be simplified as 

return this.totalProp || await this.countTotalItems()

File: src/services/pagination/LengthAwarePaginator.js (hunk 6 above)
Comment (jpaylaga) [reply src/services/pagination/LengthAwarePaginator.js:17]: Yep. I actually just realized that a few days ago. I can refactor it though, if it is still possible at this point.

File: src/services/pagination/LengthAwarePaginator.js (hunk 7 above)
Comment (jpaylaga) [reply src/services/pagination/LengthAwarePaginator.js:18]: Because there is an method called `countTotalItems()` if total params is `null`.

File: src/services/pagination/LengthAwarePaginator.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    return this.totalProp;
+  }
+
+  async toObject() {
+    const obj = await super.toObject();
+    const { items } = obj;
+    delete obj.items;
+
+    return {
+      ...obj,
```
Comment (chocnut) [inline src/services/pagination/LengthAwarePaginator.js:58]: I think it will be the same outcome since you add the 'items' field here, it was spread here, why do we need to delete it if we're returning it again?

File: src/services/pagination/LengthAwarePaginator.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+  // They act as protected methods.
+
+  /**
+   * Count total items with basic implementation.
+   *
+   * @returns {Promise<number>}
+   */
+  async countTotalItems() {
+    const resp = await this.dbProp.select(this.sqlProp, this.sqlParamsProp);
```
Comment (chocnut) [inline src/services/pagination/LengthAwarePaginator.js:73]: I don't know the exact function of this but it looks like an SQL query to me, will it be possible just to return the count output from SQL rather than this?

File: src/services/pagination/LengthAwarePaginator.js (hunk 9 above)
Comment (jpaylaga) [reply src/services/pagination/LengthAwarePaginator.js:58]: Makes sense. Thanks.

File: src/services/pagination/Paginator.js [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+   * Count all items in the current page.
+   *
+   * @returns {number}
+   */
+  async count() {
+    if (this.response.length <= 0) {
+      await this.executeQuery();
+    }
+
+    return this.response.length;
```
Comment (chocnut) [inline src/services/pagination/Paginator.js:55]: return this.response.length || await this.executeQuery();

File: src/services/pagination/LengthAwarePaginator.js (hunk 10 above)
Comment (jpaylaga) [reply src/services/pagination/LengthAwarePaginator.js:73]: Because it would be hard to parse the supplied SQL query string to include the `count()` function in it, as that query will be used to get the actual data as well. The value of the `this.sqlProp` in this context would be similar to `SELECT table_1, table_2 FROM sample_table WHERE id = 1`.

Comment (jpaylaga) [reply src/services/pagination/LengthAwarePaginator.js:73]: And I don't see a method in the `AuroraDBService` that would return the count of a query.

File: src/services/pagination/LengthAwarePaginator.js (hunk 8 above)
Comment (jpaylaga) [reply src/services/pagination/LengthAwarePaginator.js:46]: Thanks @chocnut

File: src/services/pagination/Paginator.js (hunk 11 above)
Comment (jpaylaga) [reply src/services/pagination/Paginator.js:55]: Thanks @chocnut

File: src/utils/isEmail.js (hunk 3 above)
Comment (mosufy) [reply src/utils/isEmail.js:7]: Nope, we are not using typescript. TypeScript was used on a different project, not the lesgo-framework. Eventually we'll get there

File: src/services/__tests__/LengthAwarePaginator.spec.js (hunk 4 above)
Comment (mosufy) [reply src/services/__tests__/LengthAwarePaginator.spec.js:22]: This is a simple paginator for now, to allow our current AuroraDbService to integrate with this Paginator class. We'll definitely find better ways to achieve this in later iterations

File: src/services/pagination/LengthAwarePaginator.js (hunk 6 above)
Comment (mosufy) [reply src/services/pagination/LengthAwarePaginator.js:17]: My suggestion would be to have perPage, currentPage, total in the additional params object since these are optional

File: src/services/pagination/LengthAwarePaginator.js (hunk 7 above)
Comment (mosufy) [reply src/services/pagination/LengthAwarePaginator.js:18]: This was the same issue raised and resolved in an earlier call out. Joshua will be doing the necessary corrections

File: src/utils/__mocks__/db.js (hunk 5 above)
Comment (mosufy) [reply src/utils/__mocks__/db.js:52]: Not intending to run actual db queries on unit tests for now. Everything is being mocked

File: src/utils/isDecimal.js (hunk 2 above)
Comment (mosufy) [reply src/utils/isDecimal.js]: Done thanks!

**Review by jpaylaga — APPROVED** [review]

**Review by jpaylaga — APPROVED** [review]

**Review by jpaylaga — APPROVED** [review]
I approve this PR.
